### ISSUE TITLE
Writing quantity with user-defined unit to fits table breaks if unit is registered

### ISSUE THREAD

**Issue by mhvk** [astropy/astropy#8897]
From #8883: one can write a user defined unit to a fits table, and generally this works though one gets an appropriate warning:
```
import numpy as np
from astropy import units as u
from astropy.table import QTable

bandpass_sol_lum = u.def_unit('bandpass_sol_lum') 
lum_array = np.ones((5,5)) * bandpass_sol_lum
qt = QTable()
qt['Luminosity'] = lum_array 
qt.write('output_lum.fits',overwrite=True,format='fits')
# WARNING: The unit 'bandpass_sol_lum' could not be saved to FITS format [astropy.io.fits.convenience]

u.add_enabled_units(bandpass_sol_lum)
QTable.read('output_lum.fits', format='fits')

<QTable length=5>
 Luminosity [5] 
bandpass_sol_lum
    float64     
----------------
      1.0 .. 1.0
      1.0 .. 1.0
      1.0 .. 1.0
      1.0 .. 1.0
      1.0 .. 1.0
```
So far, so good. But when one then tries to write the fits file again, it fails:
```
qt.write('output_lum.fits',overwrite=True,format='fits')
WARNING: The unit 'bandpass_sol_lum' could not be saved to FITS format [astropy.io.fits.convenience]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-c1bb5ecc6cf2> in <module>()
----> 1 qt.write('output_lum.fits',overwrite=True,format='fits')

/data/mhvk/venv/numpy-dev/lib/python3.7/site-packages/numpy-1.17.0.dev0+0dcd300-py3.7-linux-x86_64.egg/astropy/table/connect.py in __call__(self, *args, **kwargs)
    112         instance = self._instance
    113         with serialize_method_as(instance, serialize_method):
--> 114             registry.write(instance, *args, **kwargs)

/data/mhvk/venv/numpy-dev/lib/python3.7/site-packages/numpy-1.17.0.dev0+0dcd300-py3.7-linux-x86_64.egg/astropy/io/registry.py in write(data, format, *args, **kwargs)
    564 
    565     writer = get_writer(format, data.__class__)
--> 566     writer(data, *args, **kwargs)
    567 
    568 

/data/mhvk/venv/numpy-dev/lib/python3.7/site-packages/numpy-1.17.0.dev0+0dcd300-py3.7-linux-x86_64.egg/astropy/io/fits/connect.py in write_table_fits(input, output, overwrite)
    385     input = _encode_mixins(input)
    386 
--> 387     table_hdu = table_to_hdu(input, character_as_bytes=True)
    388 
    389     # Check if output file already exists

/data/mhvk/venv/numpy-dev/lib/python3.7/site-packages/numpy-1.17.0.dev0+0dcd300-py3.7-linux-x86_64.egg/astropy/io/fits/convenience.py in table_to_hdu(table, character_as_bytes)
    530 
    531             # Try creating a Unit to issue a warning if the unit is not FITS compliant
--> 532             Unit(col.unit, format='fits', parse_strict='warn')
    533 
    534     # Column-specific override keywords for coordinate columns

/data/mhvk/venv/numpy-dev/lib/python3.7/site-packages/numpy-1.17.0.dev0+0dcd300-py3.7-linux-x86_64.egg/astropy/units/core.py in __call__(self, s, represents, format, namespace, doc, parse_strict)
   1843 
   1844         elif s is None:
-> 1845             raise TypeError("None is not a valid Unit")
   1846 
   1847         else:

TypeError: None is not a valid Unit
```
Anyone any ideas? (cc @taldcroft, @astrofrog, @MSeifert04)

**Comment by pllim** [comment]
What if you start a new session, but move `u.add_enabled_units(bandpass_sol_lum)` before the first `write()` call? Just curious.

**Comment by mhvk** [comment]
Yes, tried that: it breaks the same way. Looking at the traceback, somehow the column looses its unit altogether, which is very weird. But I haven't had time to look at the actual code.

**Comment by zpace** [comment]
The plot thickens: when I change to `format=ascii.ecsv`, this table writes and reads seamlessly!

**Comment by mhvk** [comment]
@zpace - so, it is indeed fits table column related, good to know! Let's be sure to add a test also for `ecsv` so that it won't regress.

**Comment by mhvk** [comment]
Following up on @zpace's note that `ecsv` works, I made a quick PR adding a test to ensure that continues to be the case: #9013

**Comment by mhvk** [comment]
Possible fix for the fits case in #9015
